### PR TITLE
Homework 03

### PR DESCRIPTION
**Pull request by Fle9ma**
# Assignment
- [x] endpoint for adding a car
- [x] endpoint for adding a car check-up
- [x] endpoint for fetching car details along with car check-ups
  - [x] this endpoint should indicate if a new check-up is necessary
- [x] endpoint that fetches analytics showing how many check-ups were made for all manufacturers
  - [x] can be presented as a key-value pair, where key is a manufacturer, and value is a count of check-ups
 
Since we still haven’t talked about database, you can continue using in-memory store for this homework. To store cars and car check-ups in memory, you can use the same approach as in the previous homework.

### CODE REVIEW

**Comment by ivanmiklec** [comment]
Proletio sam kroz PR na brzinu, sutra još pogledam detaljnije, ali za sad nemam nekih primjedbi, super izgleda. Sviđa mi se odvajanje repozitorija/servisa/kontrolera, slojevi su potpuno odvojeni i "narušavaju" jedan drugog (repozitorij vadi podatke iz baze, servis odrađuje logiku, a kontroler sve to da klijentu).

Testovi rade i paketi(direktoriji) su jako organizirani 💪

File: academy_project/src/test/kotlin/com/infinum/course/car/checkup/car/controller/CarControllerTest.kt [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            .andExpect(MockMvcResultMatchers.jsonPath("$.car.id").value("99164f8e-e058-4815-acbf-4c79cd646867"))
+            .andExpect(MockMvcResultMatchers.jsonPath("$.car.dateAdded").isNotEmpty)
+            .andExpect(MockMvcResultMatchers.jsonPath("$.car.manufacturer").value("BMW"))
+            .andExpect(MockMvcResultMatchers.jsonPath("$.car.model").value("M3"))
+            .andExpect(MockMvcResultMatchers.jsonPath("$.car.productionYear").value(2019))
+            .andExpect(MockMvcResultMatchers.jsonPath("$.car.vin").value("12345678901234567"))
+            .andExpect(MockMvcResultMatchers.jsonPath("$.car.checkUps").isEmpty)
+            .andExpect(MockMvcResultMatchers.jsonPath("$.isCarCheckUpNecessary").value(false))
+            .andExpect(MockMvcResultMatchers.status().isOk)
+    }
```
Comment (ivanmiklec) [inline academy_project/src/test/kotlin/com/infinum/course/car/checkup/car/controller/CarControllerTest.kt:91]: Ne znam koliko si istraživao, ali možeš i testirat slučaj gdje bacaš exception klijentu pa provjerit kakav response ćeš vratit klijentu, npr. :
```
 @Test
    fun getCarById_CarDoesNotExist_ShouldReturnNotFound() {
        Mockito.`when`(carService.findCarById("99164f8e-e058-4815-acbf-4c79cd646867"))
            .thenAnswer{throw CarNotFoundException("99164f8e-e058-4815-acbf-4c79cd646867")}
        mockMvc.perform(
            MockMvcRequestBuilders
                .get("/car/find/99164f8e-e058-4815-acbf-4c79cd646867")
                .contentType(MediaType.APPLICATION_JSON)
        )
            .andExpect(MockMvcResultMatchers.jsonPath("$.message").value("Car with id 99164f8e-e058-4815-acbf-4c79cd646867 does not exist"))
            .andExpect(MockMvcResultMatchers.jsonPath("$.status").value(404))
            .andExpect(MockMvcResultMatchers.status().isNotFound)
    }
    ```

File: academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/controller/CarController.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import com.infinum.course.car.checkup.exception.NotCarException
+import org.springframework.http.HttpStatus
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.*
+
+@RestController
+@RequestMapping("/car")
+class CarController(private val carService: CarService) {
+
+    @PostMapping("/add")
```
[the remark this answers is not in the document]
Comment (Fle9ma) [reply academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/controller/CarController.kt:16]: Hvala na info 😊

File: academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/controller/CarController.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        val car = carService.findCarById(id)
+        car.checkUps.sortByDescending { it.performedAt }
+        val data = CarDetailsData(car = car, isCarCheckUpNecessary = carService.isCheckUpNecessary(id))
+        return ResponseEntity.ok(data)
+    }
+
+    @ExceptionHandler
+    fun handleNotCarException(exception: NotCarException): ResponseEntity<ErrorMessage> {
+        val errorMessage = ErrorMessage(exception.message, HttpStatus.NOT_FOUND.value())
+        return ResponseEntity(errorMessage, HttpStatus.BAD_REQUEST)
```
[the remark this answers is not in the document]
Comment (Fle9ma) [reply academy_project/src/main/kotlin/com/infinum/course/car/checkup/car/controller/CarController.kt:35]: Naravno, ima smisla, ty 🙌

File: academy_project/src/test/kotlin/com/infinum/course/car/checkup/car/controller/CarControllerTest.kt (hunk 1 above)
Comment (Fle9ma) [reply academy_project/src/test/kotlin/com/infinum/course/car/checkup/car/controller/CarControllerTest.kt:91]: Odlicno za znat, ty 👍

**Pull request merged**